Assets/Scripts/ThirdPersonController.cs

```diff
@@ -0,0 +1,119 @@
+    private Quaternion nextRotation;
+    private Vector3 nextPosition;
+    private float verticalSpeed;
+
+    public GameObject followTransform;
+    [Range (0f, 10f)] public float MoveSpeed = 6f;
+    [Range (0f, 10f)] public float LookSpeed = 5f;
+    [Range (0f, 10f)] public float Gravity = 5f;
+    [Range (0f, 10f)] public float PushPower = 5f;
+    public bool InvertY = false;
+
+    public void OnMove (InputAction.CallbackContext context) {
+        _moveInput = context.ReadValue<Vector2> ();
+    }
+
+    public void OnLook (InputAction.CallbackContext context) {
+        _lookInput = context.ReadValue<Vector2> ();
+    }
+
+    private void Awake () {
+        _controller = GetComponent<CharacterController> ();
+    }
+
+    private void Update () {
+        Move ();
+    }
+
+    private void Move () {
+
+        float lookSpeed = LookSpeed / 100;
+
+        //Move the player based on the X input
+        transform.rotation *= Quaternion.AngleAxis (_lookInput.x * lookSpeed, Vector3.up);
+
+        //Rotate the Follow Target transform based on the input
+        float lookInputY = InvertY ? -_lookInput.y : _lookInput.y;
+        followTransform.transform.rotation *= Quaternion.AngleAxis (_lookInput.x * lookSpeed, Vector3.up);
+        followTransform.transform.rotation *= Quaternion.AngleAxis (lookInputY * lookSpeed, Vector3.right);
+
+        Vector3 angles = followTransform.transform.localEulerAngles;
+        angles.z = 0;
+
+        float angle = followTransform.transform.localEulerAngles.x;
+
+        //Clamp the Up/Down rotation
+        if (angle > 180 && angle < 340) {
+            angles.x = 340;
+        } else if (angle < 180 && angle > 40) {
+            angles.x = 40;
+        }
+
+        followTransform.transform.localEulerAngles = angles;
+
+        nextRotation = Quaternion.Lerp (followTransform.transform.rotation, nextRotation, Time.deltaTime * lookSpeed);
+
+        if (_moveInput.x == 0 && _moveInput.y == 0) {
+            nextPosition = transform.position;
+
+            // transform.rotation = Quaternion.Euler (0, followTransform.transform.rotation.eulerAngles.y, 0);
+            // followTransform.transform.localEulerAngles = new Vector3 (angles.x, 0, 0);
+
+            return;
+        }
+
+        Vector3 position = (transform.forward * _moveInput.y * MoveSpeed) + (transform.right * _moveInput.x * MoveSpeed);
+        nextPosition = transform.position + position;
+
+        //Set the player rotation based on the look transform
+        transform.rotation = Quaternion.Euler (0, followTransform.transform.rotation.eulerAngles.y, 0);
+        //reset the y rotation of the look transform
+        followTransform.transform.localEulerAngles = new Vector3 (angles.x, 0, 0);
+
+        //Calculate vertical speed and gravity
+        if (_controller.isGrounded) {
+            verticalSpeed = 0;
+        }
+
+        verticalSpeed -= Gravity;
+        position.y = verticalSpeed;
```

Comment:
Again not a huge deal but I don't really understand this. Line 86 looks right, if it's a speed, modifying it by gravity makes sense. but line 87 implies it's a position not a speed.

Reply:
I guess verticalMovement would make more sense. I expect more factors in the future to contribute to y movement. Jumping for example.